core-rust/state-manager/src/store/gc.rs

```diff
@@ -0,0 +1,459 @@
+                                    DbSortKey(bytes("beef")) => DatabaseUpdate::Set(vec![2]),
+                                    DbSortKey(bytes("ee")) => DatabaseUpdate::Set(vec![3]),
+                                    DbSortKey(bytes("bada55")) => DatabaseUpdate::Set(vec![4]),
+                                )
+                            },
+                            64 => PartitionDatabaseUpdates::Delta {
+                                substate_updates: indexmap!(
+                                    DbSortKey(bytes("af")) => DatabaseUpdate::Set(vec![0]),
+                                )
+                            }
+                        )
+                    }
+                ),
+            },
+        );
+
+        // - Add some more substates to the partition 07 (so that JMT contains nodes of different versions!)
+        put_at_next_version(
+            &mut tree_store,
+            Some(1),
+            &DatabaseUpdates {
+                node_updates: indexmap!(
+                    bytes("c0ffee") => NodeDatabaseUpdates {
+                        partition_updates: indexmap!(
+                            7 => PartitionDatabaseUpdates::Delta {
+                                substate_updates: indexmap!(
+                                    DbSortKey(bytes("deafd00d")) => DatabaseUpdate::Set(vec![5]),
+                                    DbSortKey(bytes("dead")) => DatabaseUpdate::Set(vec![6]),
+                                    DbSortKey(bytes("b000000000")) => DatabaseUpdate::Set(vec![7]),
+                                )
+                            },
+                        )
+                    }
+                ),
+            },
+        );
+
+        // - Delete the extremely large partition 07:
+        put_at_next_version(
+            &mut tree_store,
+            Some(2),
+            &DatabaseUpdates {
+                node_updates: indexmap!(
+                    bytes("c0ffee") => NodeDatabaseUpdates {
+                        partition_updates: indexmap!(
+                            7 => PartitionDatabaseUpdates::Reset {
+                                new_substate_values: indexmap!()
+                            }
+                        )
+                    }
+                ),
+            },
+        );
+
+        // - Expect that subtree of our partition became stale:
+        let mut stale_subtrees = tree_store
+            .stale_part_buffer
+            .iter()
+            .filter_map(|part| match part {
+                StaleTreePart::Node(_) => None,
+                StaleTreePart::Subtree(subtree_root_key) => Some(subtree_root_key),
+            })
+            .collect::<Vec<_>>();
+        assert_eq!(stale_subtrees.len(), 1);
+        let deleted_partition_root_key = stale_subtrees.remove(0);
+        // Note: "5f" is the tier separator byte - an implementation detail of our [`NestedTreeStore`].
```

Comment:
> why do we have a Tier separator byte?

We need to somehow distinguish the RocksDB row keys of the following 2 JMT Nodes:
- a Partition-tier leaf of `{ReNode(...0fe1), Partition(a3)}`,
- a Substate-tier root (internal node) of `{ReNode(...0fe1), Partition(a3)}`.

Conceptually they both could be encoded to `...0fe1a3`.
But these are 2 different Nodes, and we can only ensure they have unique RocksDB row keys when we have a separator byte between Tiers.

> And why is it 5f?

Arbitrary choice (character `_`; a separator). It could be any value. We do not rely on its value, just on its existence.

> Why is it a byte rather than a nibble?

RocksDB row keys are made of bytes; Even the JMT's `NibblePath`s are encoded to bytes before being turned into RocksDB row keys (they use an indicator of whether they are odd-length or even-length).

> How do we distinguish it from other 5f nibbles?

We don't need to distinguish this _byte_ from other nibble-pairs of a `NibblePath`. Impl-wise, this byte is only used as a part of "prefix of all keys in the nested TreeStore".

I agree it is all a bit complicated :(